ecosystem/indexer-grpc/indexer-grpc-data-service-v2/src/connection_manager.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::config::MAX_MESSAGE_SIZE;
+use aptos_indexer_grpc_utils::{system_time_to_proto, timestamp_now_proto};
+use aptos_protos::indexer::v1::{
+    grpc_manager_client::GrpcManagerClient, service_info::Info, ActiveStream, HeartbeatRequest,
+    HistoricalDataServiceInfo, LiveDataServiceInfo, ServiceInfo, StreamInfo, StreamProgress,
+    StreamProgressSampleProto,
+};
+use dashmap::DashMap;
+use rand::prelude::*;
+use std::{
+    collections::VecDeque,
+    sync::atomic::{AtomicU64, Ordering},
+    time::{Duration, SystemTime},
+};
+use tonic::{codec::CompressionEncoding, transport::channel::Channel};
+use tracing::{info, warn};
+
+pub static MAX_HEARTBEAT_RETRIES: usize = 3;
+
+static OLD_PROGRESS_SAMPLING_RATE: Duration = Duration::from_secs(60);
+static RECENT_PROGRESS_SAMPLING_RATE: Duration = Duration::from_secs(1);
+static MAX_RECENT_SAMPLES_TO_KEEP: usize = 60;
+static MAX_OLD_SAMPLES_TO_KEEP: usize = 60;
+
+#[derive(Default)]
+struct StreamProgressSamples {
+    old_samples: VecDeque<StreamProgressSample>,
+    recent_samples: VecDeque<StreamProgressSample>,
+}
+
+impl StreamProgressSamples {
+    fn new() -> Self {
+        Default::default()
+    }
+
+    fn to_proto(&self) -> Vec<StreamProgressSampleProto> {
+        self.old_samples
+            .iter()
+            .chain(self.recent_samples.iter())
+            .map(|sample| StreamProgressSampleProto {
+                timestamp: Some(system_time_to_proto(sample.timestamp)),
+                version: sample.version,
+                size_bytes: sample.size_bytes,
+            })
+            .collect()
+    }
+
+    fn maybe_add_sample(&mut self, version: u64, size_bytes: u64) {
```

Comment:
does `ringbuffer` make this simpler?

Reply:
not really, since I'm keeping two VecDeque. (the `push` doesn't return the oldest element when it's full in ringbuffer crate)